### PR TITLE
Add support for COPY TO STDOUT and COPY FROM STDIN

### LINKED ISSUE #183: Support COPY .. TO STDOUT and COPY .. FROM STDIN using streams
For initial stream support see #176. Once that is merged we can work on supporting bulk copies.

/cc @benwilson512 (sorry to the wrong ben wilson)

### PR DESCRIPTION
**Pull request by fishcakez**
Closes #183 and #202.

This uses a different approach for copying "FROM STDIN" or to the database to #202, one query is done and then items are added a continuous stream. Importantly it adds support for copying to the database using a single query and allows data chunks to contain incomplete rows when using a stream. This requires some more complex code to ensure that streaming and non-streaming versions have identical behaviour and that we support `mode: :savepoint`.

Note that if an error occurs causing the `Enum.into/2,3` to be halted the copy as a whole fails and marks the transaction as failed (unless using `mode: :savepoint`) but does not raise (hide the initial failure).

Some more test are required to check edge cases.

### CODE REVIEW

File: lib/postgrex.ex [hunk 1]
```diff
@@ -160,6 +168,9 @@ defmodule Postgrex do
     and decoding;
     * `:mode` - set to `:savepoint` to use a savepoint to rollback to before the
     prepare on error, otherwise set to `:transaction` (default: `:transaction`);
+    * `:copy_data` - Whether to add copy data as the final parameter for use
+    with `COPY .. FROM STDIN` queries, if the query is not copying to the
+    database then the data is sent but ignore (default: `false`);
```
Comment (ericmj) [inline lib/postgrex.ex]: ignoreD

**Comment by ericmj** [comment]
This seems like a better solution indeed. Great stuff as usual, I have no comments. Feel free to merge when ready and when you've squashed the commits..

**Comment by fishcakez** [comment]
Is it ok to keep the copy to and from separate commits as they are large and reasonably independent?

**Comment by ericmj** [comment]
Of course, I just skimmed over the commit titles and thought they were the same.
